symengine/polynomial.cpp

```diff
@@ -140,23 +140,37 @@ integer_class UnivariateIntPolynomial::max_coef() const {
 }
 
 integer_class UnivariateIntPolynomial::eval(const integer_class &x) const {
-    //TODO: Use Horner's Scheme
-    integer_class ans(0), temp;
-    for (const auto &p : dict_) {
-        mp_pow_ui(temp, x, p.first);
-        ans += p.second * temp;
+
+    unsigned int last_deg = dict_.rbegin()->first;
+    integer_class result(0), x_pow;
+
+    for (auto it = dict_.rbegin(); it != dict_.rend(); ++it) {
+
+        mp_pow_ui(x_pow, x, last_deg - (*it).first);
```

Comment:
No, x_pow from the last iteration may even be bigger than needed in this iteration.
eg. `x^8 + x^2 + x`
`x_pow iteration 1 : x^6`
`x_pow iteration 2 : x`